src/test/java/codeu/controller/RegisterServletTest.java

```diff
@@ -1,35 +1,77 @@
 package codeu.controller;
 
 import java.io.IOException;
+import java.io.StringWriter;
+import java.io.PrintWriter;
 import javax.servlet.RequestDispatcher;
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import org.mockito.ArgumentCaptor;
+import org.junit.Assert;
 import org.junit.Before;
 import org.junit.Test;
 import org.mockito.Mockito;
+import codeu.model.data.User;
+import codeu.model.store.basic.UserStore;
 
 public class RegisterServletTest {
 
- private RegisterServlet registerServlet;
- private HttpServletRequest mockRequest;
- private HttpServletResponse mockResponse;
- private RequestDispatcher mockRequestDispatcher;
-
- @Before
- public void setup() {
-   registerServlet = new RegisterServlet();
-   mockRequest = Mockito.mock(HttpServletRequest.class);
-   mockResponse = Mockito.mock(HttpServletResponse.class);
-   mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
-   Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/register.jsp"))
-       .thenReturn(mockRequestDispatcher);
- }
-
- @Test
- public void testDoGet() throws IOException, ServletException {
-   registerServlet.doGet(mockRequest, mockResponse);
-
-   Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
- }
+  private RegisterServlet registerServlet;
+  private HttpServletRequest mockRequest;
+  private HttpServletResponse mockResponse;
+  private RequestDispatcher mockRequestDispatcher;
+
+  @Before
+  public void setup() {
+    registerServlet = new RegisterServlet();
+    mockRequest = Mockito.mock(HttpServletRequest.class);
+    mockResponse = Mockito.mock(HttpServletResponse.class);
+    mockRequestDispatcher = Mockito.mock(RequestDispatcher.class);
+    Mockito.when(mockRequest.getRequestDispatcher("/WEB-INF/view/register.jsp")).thenReturn(mockRequestDispatcher);
+  }
+
+  @Test
+  public void testDoGet() throws IOException, ServletException {
+    registerServlet.doGet(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testDoPost_BadUsername() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("bad !@#$% username");
+
+    registerServlet.doPost(mockRequest, mockResponse);
+
+    Mockito.verify(mockRequest).setAttribute("error", "Please enter only letters, numbers, and spaces.");
+    Mockito.verify(mockRequestDispatcher).forward(mockRequest, mockResponse);
+  }
+
+  @Test
+  public void testDoPost_NewUser() throws IOException, ServletException {
+    Mockito.when(mockRequest.getParameter("username")).thenReturn("Testertest");
+    Mockito.when(mockRequest.getParameter("password")).thenReturn("password");
+
+    StringWriter strWriter = new StringWriter();
+    PrintWriter priWriter = new PrintWriter(strWriter);
```

Comment:
Nit: I would name this printWriter instead, since pri isn't a super standard abbreviation, and spelling it out completely makes it completely clear what it means (for only two extra characters!) :)

You could also inline these writers into the .thenReturn, since they aren't used anywhere else.

OR you could return a mock here so that you can test that the name and password are set correctly in the response.

Reply:
I gotta say i'm thoroughly confused sorry x_x. I'm very unacquainted with Mockito and i really can't tell what it means to test whether it was set correctly.. Do you mean a Mockito.verify(mockRequest).getParameter ?
See if the parameter function was run once?

I have done the name changes, and i don't understand inlining the writers either, they're called at the end of the Registerservlet as two separate calls so i don't think i can combine them

Reply:
Oh nevermind, i understand now! 
And i did it! haha! I think i understand Mockito a lot better now

Reply:
Awesome! :+1: :tada: This is great because it really verifies the response contents.

Mockito can be a kind of tricky tool, so please hit me up on chat if you have any questions about it!

BTW, all I meant by "inlining" was something like `Mockito.when(mockResponse.getWriter()).thenReturn(new PrintWriter(new StringWriter()));` When it wasn't used in the rest of the test, it didn't need a variable there.